### PR TITLE
security: upgrade uft_ca to PyO3 0.29

### PR DESCRIPTION
**Pull request by Goldislops**
## Phase 2B-5B-2D — bounded PyO3 0.20 → 0.29 security migration (list[int] contract preserved)

Upgrades the maintained CA-kernel extension `crates/uft_ca` from PyO3 **0.20 → 0.29**, closing three Dependabot advisories:
- **GHSA-36hh-v3qg-5jq4** (HIGH) — OOB read in `PyList`/`PyTuple` `nth`/`nth_back`
- **GHSA-chgr-c6px-7xpp** (MEDIUM) — missing `Sync` bound on `PyCFunction::new_closure`
- **GHSA-pph8-gcv7-4qj5** (LOW) — buffer overflow in `PyString::from_object`

The 2B-5B-2C preflight found none of these APIs reachable in `uft_ca` (it crosses only Rust-native types; local-only, no untrusted Python input). The fix is performed anyway because it is inexpensive and removes the vulnerable dependency versions.

### What happened (the gate earned its keep)
1. The initial compile-only migration (`Cargo.toml` 0.29 + `Bound`/`gil_used = true` module signature) passed local Rust gates **but CI `verify-python` failed**: PyO3 ≥0.23 maps Rust `Vec<u8>` returns to Python **`bytes`** (was `list[int]`), breaking the maintained state-contract tests (`assert states == [3,3,3,2]` → `b'\x03\x03\x03\x02'`; `'bytes' object does not support item assignment`).
2. **Decision: preserve the established mutable `list[int]` contract** (not adopt `bytes`, not change tests) via explicit boundary conversion.

### Changes (exactly 3 tracked files)
- `crates/uft_ca/Cargo.toml`: pyo3 `0.20` → `0.29` (`extension-module`/`optional` unchanged).
- `crates/uft_ca/src/lib.rs`: `#[pymodule(gil_used = true)]` + `&Bound<'_, PyModule>` signature; a feature-gated `py_u8_list(py, Vec<u8>) -> PyResult<Bound<PyList>>` helper (`PyList::new`); applied at the **affected Python-exposed state-vector return sites only** — `GraphLattice.step`/`get_states`; `MultiStateGraphLattice.step`/`get_states`/`step_n`/`step_par`/`step_par_n`/`step_auto`/`step_auto_n`/`snapshot`; and the two `#[pyfn]` `step_lattice_py`/`create_graph`. Each takes a hidden `Python<'py>` token and returns `PyResult<Bound<'py, PyList>>`, wrapping the kernel's **unchanged** `Vec<u8>` result.
- `AGENT_HANDOFF.md`: one Session Log entry (revised in place).

**Internal Rust `Vec<u8>` APIs and the CA kernel are untouched** (boundary-only). `Vec<u8>` **inputs** (`new`/`set_states`), scalars, `HashMap` and `Vec<usize>` returns are unaffected. `gil_used = true` remains — **free-threaded support is not claimed**. The five `#[pyfn]` are deprecated-but-working under 0.29 and **intentionally not modernized** (Gemini's "removed in 0.21" review was corrected with the compile/CI evidence and resolved).

### Validation
- ✅ **CI `verify-python`** (Python 3.12 + maturin + full `pytest tests/`, incl. sequential/parallel-equivalence + memory-determinism fences): **PASS**.
- ✅ `cargo check --features python`: 0 errors / 7 warnings (5 `#[pyfn]` deprecations + 2 pre-existing).
- ✅ `cargo test --release`: 86 passed.
- ⚠️ `cargo test --release --features python`: Windows `extension-module` test-binary link limitation (LNK1104) — recorded, **not** worked around (CI's maturin build is the proper gate).
- No tests were weakened or changed. No `Cargo.lock` added.

### Scope
- The 3 pyo3 alerts remain open (not dismissed) and will clear once the fix reaches `main` and GitHub rescans.
- The 2 Vite alerts (frontends) and the Dependabot version-update config (2B-5B-2B) are out of scope.
- **No auto-merge.** Draft — awaits Jack review + Kevin authorization.

### CODE REVIEW

**Comment by Goldislops** [comment]
## ⚠️ CI `verify-python` fails — do not merge as-is. Blocked pending a public-API decision.

The bounded migration **compiles** cleanly (`cargo check --features python`: 0 errors; `cargo test --release [--features python]`: 86 passed), but CI's Python-3.12 + maturin + `pytest tests/` gate caught a **runtime behavioral regression**.

**Root cause:** PyO3 ≥0.23 changed the default conversion of Rust `Vec<u8>` from Python `list[int]` to **`bytes`**. Every `uft_ca` Python-facing function returns `Vec<u8>` (cell states), so under 0.29 they now return `bytes` instead of `list[int]`, breaking the maintained tests that encode the public-API contract, e.g.:
- `assert states == [3, 3, 3, 2]` → got `b'\x03\x03\x03\x02'` (`test_fractal_topologies` / `test_multi_state_stepping` / `test_graph_stepping`)
- `'bytes' object does not support item assignment` (callers mutate the returned states as a list)

Per this arc's scope rules I **stopped before broad rewriting and did not change the tests**.

**Decision needed (public Python API contract):**
1. **Preserve `list[int]`** — convert the `Vec<u8>` returns to a list-mapping type (collect to `Vec<i64>` / build a `PyList`) across the Python-facing functions. Keeps the contract and the tests unchanged, but it's a surface-wide edit (every state-returning method + the 5 `#[pyfn]`), beyond a one-signature-line bump. *(Recommended, as a slightly-widened bounded arc.)*
2. **Adopt `bytes`** — accept the new return type and update downstream callers + tests. A public-API change that ripples beyond this crate; needs explicit authorization.

The direction is correct (removes the vulnerable pyo3 versions) but it needs the list-preservation pass before it can go green. The 3 pyo3 alerts remain open (not dismissed). **Not for merge.**

File: crates/uft_ca/src/lib.rs [hunk 1]
```diff
@@ -1642,8 +1642,8 @@ impl MultiStateGraphLattice {
 }
 
 #[cfg(feature = "python")]
-#[pymodule]
-fn uft_ca(_py: Python, m: &PyModule) -> PyResult<()> {
+#[pymodule(gil_used = true)]
+fn uft_ca(_py: Python<'_>, m: &Bound<'_, PyModule>) -> PyResult<()> {
```
[the remark this answers is not in the document]
Comment (Goldislops) [reply crates/uft_ca/src/lib.rs:1657]: Empirically corrected: #[pyfn(m)] is DEPRECATED, not removed, in PyO3 0.29. This branch builds and runs against 0.29 — cargo check --features python = 0 errors (the 5 PYFN_ATTRIBUTE messages are deprecation *warnings*, not a missing-macro error), the CI maturin build succeeded, and pytest tests/ ran to completion (now green after the Vec<u8> to list[int] boundary fix). So the claim that #[pyfn] was 'completely removed in 0.21' and prevents compilation is not borne out here. The five #[pyfn(m)] registrations are intentionally retained; modernization to #[pyfunction]/wrap_pyfunction! is out of scope for this bounded security migration. Resolving on the strength of the green build.

**Comment by Goldislops** [comment]
## ✅ Blocker resolved — CI `verify-python` is green on `eec1288`

The PyO3 ≥0.23 `Vec<u8>` → `bytes` regression is fixed by explicit `PyList::new` boundary conversion at the **affected Python-exposed state-vector return sites** — correcting my earlier overstatement that *every* exported function returns `Vec<u8>`: only the state-vector returns were affected; `Vec<u8>` inputs, scalar returns, `HashMap` returns, and `Vec<usize>` returns were not.

The mutable `list[int]` public contract is preserved; the maintained tests (including the determinism/equivalence + memory fences) pass unchanged. The CA kernel and internal Rust APIs are untouched. The original blocking comment is left above as historical evidence.
